libp2p/switch.nim

```diff
@@ -209,10 +209,9 @@ proc start*(s: Switch): Future[seq[Future[void]]] {.async, gcsafe.} =
   for t in s.transports: # for each transport
     for i, a in s.peerInfo.addrs:
       if t.handles(a): # check if it handles the multiaddr
-        var server = t.start(a)
+        await t.start(a)
```

Comment:
this will raise an exception if start fails, unlike the previous code which would keep going - is that intended?

Reply:
This is [how nimbus](https://github.com/status-im/nimbus-eth2/blob/07a1c5716b707e7244eb312bed68c92c1fb9508d/beacon_chain/networking/eth2_network.nim#L1175-L1179) expects it

But I can change it, I can see cases were you don't mind about loosing one transport

Reply:
not quite - if you read the code carefully, you'll see that the previous code _never raises_ an exception - that means that nimbus will never catch one either.

In fact, the way the code is currently written, exceptions raised during `start` will silently be ignored by nimbus (which is wrong) - if an exception happens, it will be stored in one of the returned futures _inside_ the sequence, but nobody ever checks those - this needs fixing both here and in nimbus

Reply:
> By the way, I never understood why we returned the futures of the transport starting? I would remove it, but it would break API. So returning an empty seq

that's also the answer to why the futures are returned: if one of the accepts fails, the calling code can in theory figure out why and which connection failed - this is of course an extremely poor API to do so (because it requires knowledge about iteration order which is an implementation detail, but nonetheless).

It's a classic example of why we've moved on from exceptions in general: the semantics are hard to reason about and it's easy to miss where errors happen and why.

Reply:
> not quite - if you read the code carefully, you'll see that the previous code _never raises_ an exception - that means that nimbus will never catch one either.
> 
> In fact, the way the code is currently written, exceptions raised during `start` will silently be ignored by nimbus (which is wrong) - if an exception happens, it will be stored in one of the returned futures _inside_ the sequence, but nobody ever checks those - this needs fixing both here and in nimbus

Yes, that's why I said that nimbus _expected_ for libp2p to throw, but indeed, libp2p doesn't throw :)

I think it would be better to return the number of transport that failed to start, and then let the application either quit immediately, or check inside `switch.transports` which one failed to start if they don't mind loosing a few ones